### ISSUE TITLE
Randomly ocurring System.AccessViolationException freezing entire screen

### ISSUE THREAD

**Issue by GAMINGNOOBdev** [ppy/osu-framework#5615]
I've been using osu-!framework for one of my personal game projects and after a while it randomly started to throw me AccessViolationExceptions. When I build and run, after some time in execution it randomly freezes my entire screen and turns it black and when the screen comes back to normal this error message is on the console:

```
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at osu.Framework.Graphics.OpenGL.Buffers.GLVertexBuffer`1[[osu.Framework.Graphics.Rendering.Vertices.TexturedVertex2D, osu.Framework, Version=2022.1213.0.0, Culture=neutral, PublicKeyToken=null]].DrawRange(Int32, Int32)
   at osu.Framework.Graphics.OpenGL.Batches.GLVertexBatch`1[[osu.Framework.Graphics.Rendering.Vertices.TexturedVertex2D, osu.Framework, Version=2022.1213.0.0, Culture=neutral, PublicKeyToken=null]].Draw()
   at osu.Framework.Graphics.Rendering.Renderer.FlushCurrentBatch(System.Nullable`1<osu.Framework.Graphics.Rendering.FlushBatchSource>)
   at osu.Framework.Graphics.Rendering.Renderer.setMaskingInfo(osu.Framework.Graphics.Rendering.MaskingInfo, Boolean, Boolean)
   at osu.Framework.Graphics.Rendering.Renderer.PopMaskingInfo()
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.DrawChildrenOpaqueInteriors(osu.Framework.Graphics.Rendering.IRenderer, osu.Framework.Graphics.Rendering.DepthValue)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.DrawOpaqueInteriorSubTree(osu.Framework.Graphics.Rendering.IRenderer, osu.Framework.Graphics.Rendering.DepthValue)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.DrawChildrenOpaqueInteriors(osu.Framework.Graphics.Rendering.IRenderer, osu.Framework.Graphics.Rendering.DepthValue)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.DrawOpaqueInteriorSubTree(osu.Framework.Graphics.Rendering.IRenderer, osu.Framework.Graphics.Rendering.DepthValue)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.DrawChildrenOpaqueInteriors(osu.Framework.Graphics.Rendering.IRenderer, osu.Framework.Graphics.Rendering.DepthValue)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.DrawChildrenOpaqueInteriors(osu.Framework.Graphics.Rendering.IRenderer, osu.Framework.Graphics.Rendering.DepthValue)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.DrawOpaqueInteriorSubTree(osu.Framework.Graphics.Rendering.IRenderer, osu.Framework.Graphics.Rendering.DepthValue)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.DrawChildrenOpaqueInteriors(osu.Framework.Graphics.Rendering.IRenderer, osu.Framework.Graphics.Rendering.DepthValue)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.DrawOpaqueInteriorSubTree(osu.Framework.Graphics.Rendering.IRenderer, osu.Framework.Graphics.Rendering.DepthValue)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.DrawChildrenOpaqueInteriors(osu.Framework.Graphics.Rendering.IRenderer, osu.Framework.Graphics.Rendering.DepthValue)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.DrawOpaqueInteriorSubTree(osu.Framework.Graphics.Rendering.IRenderer, osu.Framework.Graphics.Rendering.DepthValue)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.DrawChildrenOpaqueInteriors(osu.Framework.Graphics.Rendering.IRenderer, osu.Framework.Graphics.Rendering.DepthValue)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.DrawOpaqueInteriorSubTree(osu.Framework.Graphics.Rendering.IRenderer, osu.Framework.Graphics.Rendering.DepthValue)
   at osu.Framework.Platform.GameHost.DrawFrame()
   at osu.Framework.Threading.GameThread.processFrame()
   at osu.Framework.Threading.GameThread.RunSingleFrame()
   at osu.Framework.Threading.GameThread.<createThread>g__runWork|66_0()
   at System.Threading.Thread+StartHelper.Callback(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Threading.Thread.StartCallback()
```

I don't know what exactly is causing the problem since the timing of the freezing is always random, sometimes even when the test browser just appeared it crashes.

**Comment by bdach** [comment]
Your guess is as good as mine. This could be a driver-level failure even.

I'd start by updating graphics card drivers and seeing if it helps any.

**Comment by peppy** [comment]
Are you doing anything low level in your code with GL calls? It sounds like you're potentially corrupting GL driver memory in some way, so it would be beneficial to check any code you have which is lower level than drawables.

**Comment by GAMINGNOOBdev** [comment]
I don't have any low level calls, all of the custom classes inherit CompositeDrawable and add to the internal children. This issue started occurring after I added more screens to a screen stack and tried rendering any of those. Even after reverting changes it still occurs. I've also installed the most recent driver updates and still no change, still gives me the error.
I was using a screen stack since the beginning of the project and had the main screen render the game. I've recently added a screen to push the main screen to the stack on mouse click (for testing) but it still crashed this time with a slightly different message:
```
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at osu.Framework.Graphics.OpenGL.Buffers.GLLinearBuffer`1[[osu.Framework.Graphics.Rendering.Vertices.TexturedVertex3D, osu.Framework, Version=2022.1213.0.0, Culture=neutral, PublicKeyToken=null]].Initialise()
   at osu.Framework.Graphics.OpenGL.Buffers.GLVertexBuffer`1[[osu.Framework.Graphics.Rendering.Vertices.TexturedVertex3D, osu.Framework, Version=2022.1213.0.0, Culture=neutral, PublicKeyToken=null]].Bind(Boolean)
   at osu.Framework.Graphics.OpenGL.Buffers.GLLinearBuffer`1[[osu.Framework.Graphics.Rendering.Vertices.TexturedVertex3D, osu.Framework, Version=2022.1213.0.0, Culture=neutral, PublicKeyToken=null]].Bind(Boolean)
   at osu.Framework.Graphics.OpenGL.Buffers.GLVertexBuffer`1[[osu.Framework.Graphics.Rendering.Vertices.TexturedVertex3D, osu.Framework, Version=2022.1213.0.0, Culture=neutral, PublicKeyToken=null]].UpdateRange(Int32, Int32)
   at osu.Framework.Graphics.OpenGL.Batches.GLVertexBatch`1[[osu.Framework.Graphics.Rendering.Vertices.TexturedVertex3D, osu.Framework, Version=2022.1213.0.0, Culture=neutral, PublicKeyToken=null]].Draw()
   at osu.Framework.Graphics.Rendering.Renderer.FlushCurrentBatch(System.Nullable`1<osu.Framework.Graphics.Rendering.FlushBatchSource>)
   at osu.Framework.Graphics.Rendering.Renderer.SetActiveBatch(osu.Framework.Graphics.Rendering.IVertexBatch)
   at osu.Framework.Graphics.OpenGL.Batches.GLVertexBatch`1[[osu.Framework.Graphics.Rendering.Vertices.TexturedVertex3D, osu.Framework, Version=2022.1213.0.0, Culture=neutral, PublicKeyToken=null]].Add(osu.Framework.Graphics.Rendering.Vertices.TexturedVertex3D)
   at osu.Framework.Graphics.Lines.Path+PathDrawNode.addLineQuads(osu.Framework.Graphics.Primitives.Line, osu.Framework.Graphics.Primitives.RectangleF)
   at osu.Framework.Graphics.Lines.Path+PathDrawNode.updateVertexBuffer()
   at osu.Framework.Graphics.Lines.Path+PathDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.BufferedDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Graphics.Containers.CompositeDrawable+CompositeDrawableDrawNode.Draw(osu.Framework.Graphics.Rendering.IRenderer)
   at osu.Framework.Platform.GameHost.DrawFrame()
   at osu.Framework.Threading.GameThread.processFrame()
   at osu.Framework.Threading.GameThread.RunSingleFrame()
   at osu.Framework.Threading.GameThread.<createThread>g__runWork|66_0()
   at System.Threading.Thread+StartHelper.Callback(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Threading.Thread.StartCallback()
```

Also something I forgot to mention is after the screen comes back to normal, when I have the stable version of osu open it does not display anything at all but still runs normally without the graphics, it just doesn't display the contents of the window.

**Comment by Geekume** [comment]
Do you know any way to solve the issue?

**Comment by GAMINGNOOBdev** [comment]
I tested running my project and a newly created template project on multiple computers and on linux and windows. On linux the projects run without a problem but on windows on both the computers I tested them on it seems to cause problems even when only rendering a single screen. I am getting these results in windows: (sorry for any background noises and for bad quality, had to record this on my phone)

https://user-images.githubusercontent.com/95419791/210091928-650f885b-819d-4b30-981b-3abaf5d0e728.mp4

https://user-images.githubusercontent.com/95419791/210092116-b799bde1-1dcf-4ef2-86be-3339bd8ea206.mp4

**Comment by Flutterish** [comment]
As a loose idea, in my experience this exception usually happens when a buffer which is used (for example data is buffered or it is drawn from) is not bound. Is it possible that somehow an element or vertex buffer was not bound? Another thing that might cause this is trying to set vertex layout with no vao bound, but this can only happen in OpenGL Core mode and as far as I'm aware o!f uses compatibility mode.

**Comment by peppy** [comment]
Please link the code.

**Comment by GAMINGNOOBdev** [comment]
Here is the link to the repository: https://github.com/GAMINGNOOBdev/noobOsu
I have recently tried to implement loading textures from directories instead of using the resources assembly and that is the only low level OpenGL related thing I implemented. I removed any references to the class loading the texture from external resources but even then it crashes. I asked one of my friends to build and run and it seemed to run fine for him on windows, I do not know if this is an issue on my windows installation or if this would apply to other people as well.

**Comment by peppy** [comment]
Another thing you could try is running lazer and see if you can repro the crash. If so, it's likely display driver related at your end.

**Comment by GAMINGNOOBdev** [comment]
It looks like this was a graphics driver issue for my integrated gpu. While the desktop build of lazer and the lazer test browser were running on my integrated gpu both crashed but after I changed it to not use the integrated graphics it stopped randomly freezing/crashing. I also changed the gpu on my own project and it fixed the problem. Looks like this was just a driver issue afterall.

Thanks for the help anyways.
